File: app/src/main/java/com/sample/android/elm/data/GitHubService.java
package com.sample.android.elm.data;

import io.reactivex.Scheduler;
import io.reactivex.Single;
import org.eclipse.egit.github.core.client.GitHubClient;
import org.eclipse.egit.github.core.service.UserService;

import java.util.concurrent.Callable;

public class GitHubService implements IApiService {

    private final Scheduler scheduler;
    private GitHubClient client;

    public GitHubService(Scheduler scheduler) {
        this.scheduler = scheduler;
    }

    @Override
    public Single<Boolean> login(final String login, final String pass) {
        return Single.fromCallable(new Callable<Boolean>() {
            @Override
            public Boolean call() throws Exception {
                client = new GitHubClient();
                client.setCredentials(login, pass);
                UserService userService = new UserService(client);
                return userService.getUser() != null;
            }
        }).subscribeOn(scheduler);
    }
}

File: app/src/main/java/com/sample/android/mobius/data/GitHubService.kt
package com.sample.android.mobius.data

import io.reactivex.Scheduler
import io.reactivex.Single
import org.eclipse.egit.github.core.Repository
import org.eclipse.egit.github.core.client.GitHubClient
import org.eclipse.egit.github.core.service.StargazerService
import org.eclipse.egit.github.core.service.UserService

class GitHubService(private val scheduler: Scheduler) : IApiService {
    private var client = GitHubClient()
    private var userName : String = ""
        @Synchronized get
        @Synchronized set

    override fun login(login: String, pass: String): Single<Boolean> {
        return Single.fromCallable {
            userName = login
            true
        }.subscribeOn(scheduler)
    }

    override fun getUserName(): String {
        return userName
    }

    override fun getStarredRepos(): Single<List<Repository>> {
        return Single.fromCallable {
            val stargazerService = StargazerService(client)
            stargazerService.getStarred(userName)
        }.subscribeOn(scheduler)
    }
}
